pkg/storage/ingest/pusher.go

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingest
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/cancellation"
+	"github.com/grafana/dskit/user"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+type Pusher interface {
+	Push(context.Context, *mimirpb.WriteRequest) (*mimirpb.WriteResponse, error)
+}
+
+type pusherConsumer struct {
+	p Pusher
+
+	metrics readerMetrics
+	l       log.Logger
+}
+
+type parsedRecord struct {
+	*mimirpb.WriteRequest
+	tenantID string
+	err      error
+}
+
+func newPusherConsumer(p Pusher, metrics readerMetrics, l log.Logger) *pusherConsumer {
+	return &pusherConsumer{
+		p:       p,
+		metrics: metrics,
+		l:       l,
+	}
+}
+
+func (c pusherConsumer) consume(ctx context.Context, records []record) error {
+	recC := make(chan parsedRecord)
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(cancellation.NewErrorf("done consuming records"))
+
+	go c.unmarshalRequests(ctx, records, recC)
```

Comment:
[nit] I suggest to add a comment to explain what we're doing here. Following suggestion is just an example:

```suggestion
	// We concurrently unmarshal requests and push them to the receiver.
	go c.unmarshalRequests(ctx, records, recC)
```